### PR TITLE
tests: add a test of idempotency of same txn storage

### LINKED ISSUE #1505: Write a test that verifies the idempotency of processing the same transaction more than once
The situation when Electric receives the same transaction more than once via the replication stream can arise in the following scenario. When the database cluster shuts down, all of Electric's connections to the database will close and Electric will go into the reconnection-with-backoff loop. Once Postgres comes back up, Electric will be able to resume its logical replication stream from the same point at which it stopped streaming previously thanks to the persistent replication slot it had created.

If Electric has processed a transaction (by persisting it to the shape log) but hasn't acknowledged its WAL offset to Postgres because the connection was closed, the next time Postgres starts, Electric will read the already seen transaction from the logical stream once again.

### PR DESCRIPTION
**Pull request by icehaunter**
Closes #1505

### CODE REVIEW

**Comment by balegas** [comment]
@icehaunter see https://github.com/electric-sql/electric/issues/1505#issuecomment-2283495707

**Comment by icehaunter** [comment]
@balegas This adds the aforementioned test. As to whether we should avoid writing at all in case of duplication, I think we can loop back to that question later because under current implementations it's not necessary. If our storage engine becomes actually append-log, then it might become an important idea

**Comment by balegas** [comment]
Do you mind leaving a comment in the code noting that the way we write to storage actually implies idempotency and why it is required?

I suppose this is remains true for an in-memory approach

File: packages/sync-service/test/electric/replication/shape_log_collector_test.exs [hunk 1]
```diff
@@ -288,4 +290,79 @@ defmodule Electric.Replication.ShapeLogCollectorTest do
...
+      assert [op1, op2] =
+               Storage.get_log_stream(shape_id, LogOffset.before_all(), storage)
+               |> Enum.map(&:json.decode/1)
+
+      # If we encounter & store the same transaction, log stream should be stable
+      assert :ok = ShapeLogCollector.store_transaction(txn, server)
+
+      assert [^op1, ^op2] =
+               Storage.get_log_stream(shape_id, LogOffset.before_all(), storage)
+               |> Enum.to_list()
```
Comment (alco) [inline packages/sync-service/test/electric/replication/shape_log_collector_test.exs:364]:
```suggestion
```

File: packages/sync-service/test/electric/shape_cache/storage_implementations_test.exs [hunk 2]
```diff
@@ -199,6 +199,35 @@ defmodule Electric.ShapeCache.StorageImplimentationsTest do
...
+          storage.get_log_stream(@shape_id, LogOffset.first(), LogOffset.last(), opts)
+          |> Enum.map(&:json.decode/1)
+
+        :ok = storage.append_to_log!(@shape_id, log_items, opts)
+
+        log2 =
+          storage.get_log_stream(@shape_id, LogOffset.first(), LogOffset.last(), opts)
+          |> Enum.map(&:json.decode/1)
+
+        assert log1 == log2
```
Comment (alco) [inline packages/sync-service/test/electric/shape_cache/storage_implementations_test.exs:229]: Nit: No need to decode from JSON for this assertion.

**Review by alco — APPROVED** [review]
The code looks good but I have some thoughts on whether it addresses the issue.

Right now the replication stream is only responsible for saving incoming transactions to log storage. But we've already had discussions about streaming transactions in parallel with persisting them to log storage in some cases. As the system evolves, at some point we may replace the current `transaction_received` callback passed to `ReplicationClient` with another function potentially from another module, side-stepping the test that only verifies the behaviour of ShapeLogCollector.

I think we can potentially address the duplicate transaction processing problem at the source, i.e. by having information inside ReplicationClient about which lsn it processed last. It's just that implementing this approach requires some design work first.

File: packages/sync-service/test/electric/shape_cache/storage_implementations_test.exs (hunk 2 above)
Comment (icehaunter) [reply packages/sync-service/test/electric/shape_cache/storage_implementations_test.exs:229]: True, but the comparison error will be better

**Comment by balegas** [comment]
I see this being merged and the original issue being closed, but I feel that the PR misses the point of the issue.
Ideally the test should cover the any implementation rather than a specific one, but if it doesn't we need to document the assumption, because this implicit property is required for the correctness of the system. I think that is helpful that the test covers any changes that people might do to the current implementation, but I feel that documentation is needed for someone that is making a new storage implementation. 

I don't think we need to do the engineering ahead of switching implementations and we might actually touch that code when we do https://github.com/electric-sql/electric/issues/1506. So, I'm happy addressing the issue at the source there. However, I strongly think that we need to document in code the implicit property.

**Comment by alco** [comment]
I have filed a new issue specifically for addressing the storage of replication state for the purpose of detecting repeat transactions without involving the shape storage subsystem - https://github.com/electric-sql/electric/issues/1515.

**Comment by icehaunter** [comment]
@balegas I've added a storage-level test explicitly for this reason, so that it's less tied to the shape log collector implementation. If someone adds an append-only storage, I would assume they would look at the failing test and question the assumption explicitly stated there. Do you feel like that's not enough?

**Pull request merged**